### PR TITLE
Fix memory leak loading malformed `SawyerChunk`

### PR DESCRIPTION
**Pull request by john-h-kastner**
A temporary buffer was not free'd after failing to parse in `SawyerChunkReader::ReadChunkTrack`. Fix this following the pattern used in `SawyerChunkReader::ReadChunk` by wrapping the relevant code in a `try` block with `FreeLargeTempBuffer` called when an exception is caught.

As an alternative fix, the buffer could be allocated outside the outer `try` block a freed in the corresponding `catch`. This would result in less nesting, and slightly more pleasing code, but would mean the buffer is unnecessarily allocated and freed on some error paths where it is never used.

### CODE REVIEW

**Comment by janisozaur** [comment]
Do you have a sample object that triggers this?

**Comment by john-h-kastner** [comment]
> Do you have a sample object that triggers this?

An empty input stream will do it. This causes `uncompressedLength` to be 0, so the method throws an exception before ownership of the buffer is passed to `SawyerChunk` at the end of the block and the buffer is never freed. [`DecodeChunk` can throw exceptions for it's own reasons](https://github.com/OpenRCT2/OpenRCT2/blob/a0b18688ec203f22b79b1dea6af648f04c29369c/src/openrct2/rct12/SawyerChunkReader.cpp#L202) with the same effect.

**Comment by janisozaur** [comment]
Can you please add a test here: https://github.com/OpenRCT2/OpenRCT2/blob/develop/test/tests/SawyerCodingTest.cpp ? I can add it myself if you decline.

**Comment by janisozaur** [comment]
Actually, is it this test: https://github.com/OpenRCT2/OpenRCT2/blob/96be44153330956808c5227eb7785bfca91d9a37/test/tests/SawyerCodingTest.cpp#L174-L180 ?

**Review by janisozaur — APPROVED** [review]

**Comment by janisozaur** [comment]
On second thought, the tests are not necessary here as we don't have any for `ReadChunkTrack` yet and we (the team) should first add some with data that makes sense. I approve this PR, but need one more sign off.

**Comment by ZehMatt** [comment]
I'm honestly not a big fan on this, why not just use unique_ptr, the only reason for this special memory seems to be that its slow for debug builds? That is somewhat the purpose of debug builds, we get a special debug heap to be able to better track memory issues, this feels very counter intuitive.

**Comment by janisozaur** [comment]
@john-h-kastner will you fix this or should we take over?

File: src/openrct2/rct12/SawyerChunkReader.cpp [hunk 1]
```diff
@@ -302,25 +289,7 @@ size_t SawyerChunkReader::DecodeChunkRotate(void* dst, size_t dstCapacity, const
...
-#endif
-    if (buffer == nullptr)
-    {
-        throw std::runtime_error("Unable to allocate large temporary buffer.");
-    }
-    return buffer;
-}
-
-void SawyerChunkReader::FreeLargeTempBuffer(void* buffer)
+std::unique_ptr<uint8_t[]> SawyerChunkReader::AllocateLargeTempBuffer()
```
Comment (ZehMatt) [inline src/openrct2/rct12/SawyerChunkReader.cpp:292]: Do we really need this function?

Comment (john-h-kastner) [reply src/openrct2/rct12/SawyerChunkReader.cpp:292]: Strictly speaking, no you could in line any function of your choosing.

Comment (ZehMatt) [reply src/openrct2/rct12/SawyerChunkReader.cpp:292]: I don't think we require this function to just create a new unique_ptr is what I'm saying.

**Comment by janisozaur** [comment]
@john-h-kastner will you update the code to address the comment? Otherwise I can update it for you so it can get merged.

**Review by janisozaur — APPROVED** [review]

**Comment by janisozaur** [comment]
I assume ZehMatt's comments and my last commit count as a positive review.

**Pull request merged**